File: backend/app/api/routes/sales_routes/contact_api.py
from typing import Optional
from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import JSONResponse
from fastapi.encoders import jsonable_encoder

from app.dependency.sales.contact_dependency import get_contact_service
from app.dependency.auth_dependency import verify_access_token_dep
from app.models.sales.leader import Leader
from app.services.sales.contact_service import ContactService
from app.schemas.sales.contact_schema import (
    ContactCreateSchema,
    ContactUpdateSchema,
    ContactResponseSchema,
    ContactDropdownSchema,
)

contact_router = APIRouter(prefix="/api/v1/contacts", tags=["Contact API"])


@contact_router.get("/dropdown", status_code=status.HTTP_200_OK)
async def get_contacts_dropdown(
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Get active contacts list for dropdown."""
    user_role = (getattr(current_user, "role", "") or "").strip().lower().replace("_", " ")
    user_desig = (getattr(current_user, "designation", "") or "").strip().upper()
    is_executive = user_role in ["super admin", "admin"] or user_desig in ["CEO", "CFO"]
    leader_id = current_user.leader_id
    contacts = await service.get_contacts_dropdown(leader_id=leader_id, is_executive=is_executive)
    return JSONResponse(
        content=jsonable_encoder([ContactDropdownSchema.model_validate(c).model_dump() for c in contacts]),
        status_code=status.HTTP_200_OK
    )


@contact_router.post("", status_code=status.HTTP_201_CREATED, response_model=ContactResponseSchema)
async def create_contact(
    payload: ContactCreateSchema,
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Create a new contact record."""
    if not payload.created_by:
        payload.created_by = current_user.leader_id or current_user.emp_id
    contact = await service.create_contact(payload)
    return JSONResponse(
        content=jsonable_encoder(ContactResponseSchema.model_validate(contact)),
        status_code=status.HTTP_201_CREATED,
    )
from app.schemas.common.bulk_delete_schema import BulkDeleteRequest


@contact_router.delete("/bulk", status_code=status.HTTP_200_OK)
async def bulk_delete_contacts(
    payload: BulkDeleteRequest,
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Delete multiple selected contact records."""
    count = await service.bulk_delete_contacts(contact_ids=payload.ids, current_user=current_user)
    return JSONResponse(
        content={"message": f"Successfully deleted {count} contact records.", "deleted_count": count},
        status_code=status.HTTP_200_OK,
    )


@contact_router.delete("/delete-all", status_code=status.HTTP_200_OK)
async def delete_all_contacts(
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Delete all contact records (Super Admin / Executive only)."""
    count = await service.delete_all_contacts(current_user=current_user)
    return JSONResponse(
        content={"message": f"Successfully deleted all contacts ({count} records).", "deleted_count": count},
        status_code=status.HTTP_200_OK,
    )


@contact_router.get("/{contact_id}", status_code=status.HTTP_200_OK, response_model=ContactResponseSchema)
async def get_contact(
    contact_id: str,
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Fetch a contact record by contact_id."""
    user_role = (getattr(current_user, "role", "") or "").strip().lower().replace("_", " ")
    user_desig = (getattr(current_user, "designation", "") or "").strip().upper()
    is_executive = user_role in ["super admin", "admin"] or user_desig in ["CEO", "CFO"]
    contact = await service.get_contact_by_id(contact_id, is_executive=is_executive)
    return JSONResponse(
        content=jsonable_encoder(ContactResponseSchema.model_validate(contact)),
        status_code=status.HTTP_200_OK,
    )


@contact_router.get("", status_code=status.HTTP_200_OK)
async def list_contacts(
    limit: int = Query(50, ge=1, le=100),
    cursor: Optional[str] = Query(None, description="Keyset cursor: pass the last contact_id from previous response"),
    is_active: Optional[bool] = Query(None, description="Filter by active status (executives only)"),
    search: Optional[str] = Query(None, description="Search term for company, contact_name, designation, phone_no, email, country"),
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """List contacts with keyset pagination and multi-field search filtering."""
    user_role = (getattr(current_user, "role", "") or "").strip().lower().replace("_", " ")
    user_desig = (getattr(current_user, "designation", "") or "").strip().upper()
    is_executive = user_role in ["super admin", "admin"] or user_desig in ["CEO", "CFO"]
    leader_id = current_user.leader_id
    result = await service.get_all_contacts(
        limit=limit, cursor=cursor, leader_id=leader_id, is_executive=is_executive, is_active_filter=is_active, search=search
    )
    result["data"] = [
        ContactResponseSchema.model_validate(item).model_dump()
        for item in result["data"]
    ]
    return JSONResponse(content=jsonable_encoder(result), status_code=status.HTTP_200_OK)


@contact_router.patch("/{contact_id}", status_code=status.HTTP_200_OK, response_model=ContactResponseSchema)
async def update_contact(
    contact_id: str,
    payload: ContactUpdateSchema,
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Update a contact record."""
    user_role = (getattr(current_user, "role", "") or "").strip().lower().replace("_", " ")
    user_desig = (getattr(current_user, "designation", "") or "").strip().upper()
    is_executive = user_role in ["super admin", "admin"] or user_desig in ["CEO", "CFO"]
    contact = await service.update_contact(contact_id, payload, is_executive=is_executive)
    return JSONResponse(
        content=jsonable_encoder(ContactResponseSchema.model_validate(contact)),
        status_code=status.HTTP_200_OK,
    )


@contact_router.delete("/{contact_id}", status_code=status.HTTP_200_OK)
async def delete_contact(
    contact_id: str,
    service: ContactService = Depends(get_contact_service),
    current_user: Leader = Depends(verify_access_token_dep),
):
    """Delete a contact record."""
    user_role = (getattr(current_user, "role", "") or "").strip().lower().replace("_", " ")
    user_desig = (getattr(current_user, "designation", "") or "").strip().upper()
    is_executive = user_role in ["super admin", "admin"] or user_desig in ["CEO", "CFO"]
    await service.delete_contact(contact_id, is_executive=is_executive)
    return JSONResponse(
        content={"message": f"Contact {contact_id} deleted successfully"},
        status_code=status.HTTP_200_OK,
    )

File: frontend/src/api/contactApi.js
// src/api/contactApi.js
import axiosInstance from "./axiosInstance";
import { withAuthRetry } from "./authApi";

/**
 * Fetch paginated contacts from GET /api/v1/contacts
 * @param {Object} params - { limit: 50, cursor: null, is_active: true/null, search: string/null }
 */
export function getContacts(params = {}) {
  return withAuthRetry(async () => {
    const cleanParams = {};
    if (params.limit !== undefined && params.limit !== null) cleanParams.limit = params.limit;
    if (params.cursor !== undefined && params.cursor !== null) cleanParams.cursor = params.cursor;
    if (params.is_active !== undefined && params.is_active !== null) cleanParams.is_active = params.is_active;
    if (params.search && params.search.trim()) cleanParams.search = params.search.trim();
    if (params.from_date !== undefined && params.from_date !== null) {
      cleanParams.from_date = params.from_date;
    }
    if (params.to_date !== undefined && params.to_date !== null) {
      cleanParams.to_date = params.to_date;
    }

    // Force live fetch from backend if date parameters exist or skipCache is requested
    const shouldSkipCache = Boolean(params.skipCache || params.from_date || params.to_date);

    const res = await axiosInstance.get("/contacts", {
      params: cleanParams,
      skipCache: shouldSkipCache,
      headers: shouldSkipCache ? { "Cache-Control": "no-cache, no-store", "Pragma": "no-cache" } : {}
    });
    return res.data;
  });
}

/**
 * Fetch lightweight contacts dropdown: GET /api/v1/contacts/dropdown
 */
export function getContactsDropdown() {
  return withAuthRetry(async () => {
    const res = await axiosInstance.get("/contacts/dropdown");
    return res.data;
  });
}

/**
 * Fetch a single contact by contact_id: GET /api/v1/contacts/{contact_id}
 * @param {string} contactId
 */
export function getContactById(contactId) {
  return withAuthRetry(async () => {
    const res = await axiosInstance.get(`/contacts/${encodeURIComponent(contactId)}`);
    return res.data;
  });
}

/**
 * Create a new contact: POST /api/v1/contacts
 * Schema:
 * {
 *   company: string,
 *   contact_name: string,
 *   designation?: string,
 *   phone_no_1: string,
 *   phone_no_2?: string,
 *   email: string,
 *   country?: string,
 *   region?: string,
 *   is_active?: boolean
 * }
 * @param {Object} data
 */
export function createContact(data) {
  return withAuthRetry(async () => {
    const payload = {
      company: String(data.company || "").trim(),
      contact_name: String(data.contact_name || "").trim(),
      phone_no_1: String(data.phone_no_1 || "").trim(),
      email: String(data.email || "").trim(),
      is_active: data.is_active !== undefined ? Boolean(data.is_active) : true
    };
    if (data.designation !== undefined) payload.designation = data.designation ? String(data.designation).trim() : null;
    if (data.phone_no_2 !== undefined) payload.phone_no_2 = data.phone_no_2 ? String(data.phone_no_2).trim() : null;
    if (data.country !== undefined) payload.country = data.country ? String(data.country).trim() : null;
    if (data.region !== undefined) payload.region = data.region ? String(data.region).trim() : null;

    const res = await axiosInstance.post("/contacts", payload);
    return res.data;
  });
}

/**
 * Update an existing contact: PATCH /api/v1/contacts/{contact_id}
 * Sends ONLY the modified/provided fields in the payload.
 * Schema (all optional in PATCH):
 * {
 *   company?: string,
 *   contact_name?: string,
 *   designation?: string,
 *   phone_no_1?: string,
 *   phone_no_2?: string,
 *   email?: string,
 *   country?: string,
 *   region?: string,
 *   is_active?: boolean
 * }
 * @param {string} contactId
 * @param {Object} data
 */
export function updateContact(contactId, data = {}) {
  return withAuthRetry(async () => {
    const payload = {};
    if (data.company !== undefined) payload.company = String(data.company).trim();
    if (data.contact_name !== undefined) payload.contact_name = String(data.contact_name).trim();
    if (data.designation !== undefined) payload.designation = data.designation ? String(data.designation).trim() : null;
    if (data.phone_no_1 !== undefined) payload.phone_no_1 = String(data.phone_no_1).trim();
    if (data.phone_no_2 !== undefined) payload.phone_no_2 = data.phone_no_2 ? String(data.phone_no_2).trim() : null;
    if (data.email !== undefined) payload.email = String(data.email).trim();
    if (data.country !== undefined) payload.country = data.country ? String(data.country).trim() : null;
    if (data.region !== undefined) payload.region = data.region ? String(data.region).trim() : null;
    if (data.is_active !== undefined) payload.is_active = Boolean(data.is_active);

    const res = await axiosInstance.patch(`/contacts/${encodeURIComponent(contactId)}`, payload);
    return res.data;
  });
}

/**
 * Delete a contact: DELETE /api/v1/contacts/{contact_id}
 * @param {string} contactId
 */
export function deleteContact(contactId) {
  return withAuthRetry(async () => {
    const res = await axiosInstance.delete(`/contacts/${encodeURIComponent(contactId)}`);
    return res.data;
  });
}

/**
 * Bulk delete selected contacts: DELETE /api/v1/contacts/bulk
 * Body payload: { "ids": ["string" | number] }
 * @param {Array<number|string>|{ids: Array<number|string>}} payload
 */
export function bulkDeleteContacts(payload) {
  return withAuthRetry(async () => {
    let idsArray = [];
    if (Array.isArray(payload)) {
      idsArray = payload;
    } else if (payload && Array.isArray(payload.ids)) {
      idsArray = payload.ids;
    } else if (payload && (payload.id || payload.contact_id)) {
      idsArray = [payload.id || payload.contact_id];
    }
    const body = { ids: idsArray };
    const res = await axiosInstance.delete("/contacts/bulk", { data: body });
    return res.data;
  });
}

/**
 * Delete all contacts: DELETE /api/v1/contacts/delete-all
 */
export function deleteAllContacts() {
  return withAuthRetry(async () => {
    const res = await axiosInstance.delete("/contacts/delete-all");
    return res.data;
  });
}
